crates/ty_python_semantic/src/types/typed_dict.rs

```diff
@@ -246,14 +262,412 @@ impl<'db> TypedDictType<'db> {
         }
         constraints
     }
+
+    pub fn definition(self, db: &'db dyn Db) -> Option<Definition<'db>> {
+        match self {
+            TypedDictType::Class(defining_class) => Some(defining_class.definition(db)),
+            TypedDictType::Synthesized(_) => None,
+        }
+    }
+
+    fn params(self, db: &'db dyn Db) -> TypedDictParams {
+        match self {
+            TypedDictType::Class(defining_class) => {
+                let class_literal = defining_class.class_literal(db).0;
+                class_literal
+                    .typed_dict_params(db)
+                    .expect("must be a TypedDict")
+            }
+            TypedDictType::Synthesized(synthesized) => synthesized.params(db),
+        }
+    }
+
+    /// Return the meta-type of this `TypedDict` type.
+    pub(super) fn to_meta_class(self, db: &'db dyn Db) -> ClassType<'db> {
+        // `TypedDict` instances are instances of `dict` at runtime, but its important that we
+        // understand a more specific meta type in order to correctly handle `__getitem__`.
+        match self {
+            TypedDictType::Class(defining_class) => defining_class,
+            TypedDictType::Synthesized(_) => KnownClass::TypedDictFallback
+                .try_to_class_literal(db)
+                .map(|class| class.default_specialization(db))
+                .unwrap_or_else(|| {
+                    KnownClass::Object
+                        .try_to_class_literal(db)
+                        .map(|class| class.default_specialization(db))
+                        .expect("object class must exist")
+                }),
+        }
+    }
+
+    pub(crate) fn normalized_impl(self, db: &'db dyn Db, visitor: &NormalizedVisitor<'db>) -> Self {
+        match self {
+            TypedDictType::Class(_) => {
+                let synthesized =
+                    SynthesizedTypedDictType::new(db, self.params(db), self.items(db));
+                TypedDictType::Synthesized(synthesized.normalized_impl(db, visitor))
```

Comment:
Would it be better to inline `SynthesizedTypedDictType::normalized_impl` here instead of creating two instances here and throwing the first away?

Reply:
I think this is fine